Specification of FeedbackEntity

raphaellith opened the pull request:
This PR adds specifications for FeedbackEntity, which represents a piece of textual feedback for a response previously submitted by a student.

The procedure by which an Android tablet retrieves responses from the Windows laptop mirrors the previously defined process for retrieving materials.

## Questions to consider

The `FeedbackEntity` specified here applies to responses submitted to all question types. As the specifications for `ResponseEntity` currently stand, they include the following optional Boolean field `IsCorrect`, stating that "the Android device may assist in validating simple questions (such as Multiple Choice, True/False or simple blank filling)". However, the scope of "simple questions" have yet to be finalised. When should feedback be given through `FeedbackEntity`, and when should it be determined by the Android device itself?

nemoshu commented on docs/Validation Rules.md:
```diff
@@ -141,6 +141,22 @@ If validation rules in API Contract are in contradiction to this document, this
+    (a) `ResponseId` (UUID). The response targeted by this feedback.
+
+    (b) `Text` (String): Textual feedback.
+
```
I believe a feedback entity should also be able to support a numerical mark?

I'm considering a logical OR relationship - as long as an entity contains one of the two types of feedback it is valid.

raphaellith commented:
New push: 
1. Added optional `MaxScore` field to `QuestionEntity`
1. Added `Marks` field to `FeedbackEntity` (a valid `FeedbackEntity` must have a `Text` field, a `Marks` field, or both)
1. Feedback is only allowed for non-simple questions, where responses cannot be marked on-tablet.
1. Other related validation constraints

raphaellith commented:
My main question with this new push is item 3:

> Feedback is only allowed for non-simple questions, where responses cannot be marked on-tablet.

Currently this restricts both numerical and textual feedback to more complex questions like `WRITTEN_ANSWER`. The main idea behind this is that simple questions like multiple choice and true/false can be directly marked (numerically) on-tablet.

However, this would also outlaw any textual feedback for simple questions.

Should we remove this constraint to allow textual feedback for true/false and multiple choice questions?

raphaellith commented:
This feels potentially debatable given the use cases of this application.
@PriyaBargota @WillStephenn

WillStephenn commented on docs/API Contract.md:
```diff
@@ -309,6 +326,7 @@ This section documents explicit and implicit acknowledgement mechanisms for TCP
 | `PAIRING_REQUEST (0x20)` | `PAIRING_ACK (0x21)` | Client → Server, Server → Client |
 | `HAND_RAISED (0x11)` | `HAND_ACK (0x06)` | Client → Server, Server → Client |
 | `DISTRIBUTE_MATERIAL (0x05)` | `DISTRIBUTE_ACK (0x12)` | Server → Client, Client → Server |
+| `RETURN_FEEDBACK (0x06)` | `FEEDBACK_ACK (0x13)` | Server → Client, Client → Server |
```
This should be 0x07, 0x06 is for `'HAND_ACK`'

PriyaBargota commented:
> @PriyaBargota I have opened Issue #163 addressing the Android side changes required to fulfil these doc updates.

Great, I will work on these changes.
